### ISSUE TITLE
VM `resolveClassRef()` helper crashes when resolving a resolved method

### ISSUE THREAD

**Issue by NigelYiboYu** [eclipse/openj9#4363]
This issue is only reproducible using certain OpenJ9 builds in AOT mode. 

Using Java 8 SR5 FP27 (Jan 07, 2019 build) in AOT mode, the JIT may generate a `J9UnresolvedCallSnippet` on a resolved static method. The reason is that the AOT currently is unable to relocate direct funtion calls, ie. relocating BRASL instruction on Z means patching the instruction's binary. As a result, a call snippet is required to handle direct calls in AOT mode.

Admittedly, using a `J9UnresolvedCallSnippet` on resolved static methods is abusing this VM helper API; but one should not expect segfaults in this case.

The following is an example (reproducible using the attached files).

```
// Setting up call parameters and making a BRASL call
    LGR     GPR7,GPR2    ; LR=Reg_move
    LGR     GPR1,GPR0    ; LR=Reg_param
    LGR     GPR3,GPR6    ; LR=Reg_param
    MVHI    #683 8(GPR5),0x0005
    MVHI    #684 0(GPR5),0x0001
    ASSOCREGS
    BRASL   GPR14, 0x000003FF7A6C7E60, targetAddr=0x000003FF7B61343C (offset=0x000000000000018E)
    LGR     GPR3,GPR8    ; LR=Reg_move
    LGR     GPR2,GPR7    ; LR=Reg_move
....
....
  Snippet Label L0067:          ; Unresolved Call Snippet
    ST          GPR1,32(,GPR5)
    STG         GPR2,24(,GPR5)
    ST          GPR3,16(,GPR5)
    LARL        GPR14, *+14 <000003FF7B613458>  # Start of Data Const.
    LG          GPR_EP, 0(,GPR14)
    BCR         GPR_EP
    DC          000003FF90ECD210                # Method Address
    DC          000003FF7B6132B4                # Call Site RA
    DC          0000000000000000                # Method Pointer
    DC          000003FF581CE280                # Address Of Constant Pool
    DC          0x00000046              # Offset | Flag | CP Index
```

The BRASL would be a direct jump to target callee (resolved private static method) if this program were running in non-AOT mode. In AOT however, this BRASL jumps to the `Unresolved Call Snippet` and BCR to the `old_slow_jitResolveStaticMethod()` and eventually crashed in `resolveClassRef()`.

Note that the `Address of Constant Pool` shown in the trace is a compile-time value and will be changed to a valid CP address after relocation, which happens after the AOT-compile and before the execution of the function body. To trace relocation, run the attached `run` script to obtain a `rtlog_name` trace file. It'll contain a record of relocation similar to the following snippet:
```
 97 TR_Trampolines 000003FF844FF182
 98         size 1a type 18 flags 0 reloFlags 0
 99         inlined site index FFFFFFFFFFFFFFFF
100         constant pool 000003FF5C1CE280
101         reloLocation: from 000003FF844FF19A at 000003FF84E00470 (offset 460)
102                 applyRelocation: oldAddress 000003FF5C1CE280
103                 computeNewConstantPool: newCP 000000000BFF3A90
```

GDB trace when the VM helper crashes:

```
(gdb) where
#0  resolveClassRef (vmStruct=0xbeb2200, ramCP=0xbff3a90, cpIndex=<optimized out>, resolveFlags=32) at resolvesupport.cpp:262
#1  resolveStaticMethodRefInto (vmStruct=0xbeb2200, ramCP=0xbff3a90, cpIndex=<optimized out>, resolveFlags=32, ramCPEntry=0xbeb2308) at resolvesupport.cpp:478
#2  resolveStaticMethodRef (vmStruct=0xbeb2200, ramCP=0xbff3a90, cpIndex=70, resolveFlags=32) at resolvesupport.cpp:588
#3  old_slow_jitResolveStaticMethod (currentThread=0xbeb2200) at cnathelp.cpp:1674
#4  jitResolveStaticMethod () at /tmp/yunigel/sdk-aot-daa/jre/lib/s390x/compressedrefs/compiler/../objs/compiler/z/runtime/ValueProf.o.s:1962
#5  runCallInMethod (env=0x4668bba000, receiver=0x11869a838, clazz=0xbf6fff0, methodID=<optimized out>, args=<optimized out>) at callin.cpp:1073
#6  gpProtectedRunCallInMethod (entryArg=0x3ffa797ed68) at jnicsup.cpp:256
#7  omrsig_protect () from /tmp/yunigel/sdk-aot-daa/jre/lib/s390x/compressedrefs/libj9prt29.so
#8  gpProtectAndRun (function=0x3ffa6d32678 <gpProtectedRunCallInMethod(void*)>, env=0xbeb2200, args=0x3ffa797ed68) at jniprotect.c:78
#9  gpCheckCallin (env=0xbeb2200, receiver=<optimized out>, cls=<optimized out>, methodID=<optimized out>, args=0x3ffa797ee30) at jnicsup.cpp:439
#10 callStaticVoidMethod (env=<optimized out>, cls=<optimized out>, methodID=<optimized out>) at jnicgen.c:293
#11 JavaMain () from /tmp/yunigel/sdk-aot-daa/bin/../lib/s390x/jli/libjli.so
#12 start_thread () from /lib64/libpthread.so.0
#13 thread_start () from /lib64/libc.so.6
```
`resolveClassRef()` crashed on this line: `	if ((NULL != resolvedClass) && (J9_ARE_NO_BITS_SET(resolveFlags, J9_RESOLVE_FLAG_INSTANTIABLE) || J9_ARE_NO_BITS_SET(resolvedClass->romClass->modifiers, J9AccAbstract | J9AccInterface))) {` where `resolvedClass` is a bad pointer.

The run command and Java program to reproduce the issue are attached. Rename Test.txt to Test.java.
[run.txt](https://github.com/eclipse/openj9/files/2779469/run.txt)
[Test.txt](https://github.com/eclipse/openj9/files/2779472/Test.txt)

**Comment by fjeremic** [comment]
@gacholio / @DanHeidinga who is most knowledgeable in this area who can help @NigelYiboYu with this crash in VM resolve code? I'm not 100% convinced there is an issue on the VM side as the JIT could be passing in bad arguments, however we do have a unit test which reproduces this 100% so it should be easy to debug if we can get some help around this code.

**Comment by gacholio** [comment]
First step would be to debug compile the VM - if the cpIndex is way off, it'll be immediately obvious,

**Comment by DanHeidinga** [comment]
@NigelYiboYu @fjeremic Does this fail on other platforms as well?

**Comment by gacholio** [comment]
I suspect the cpIndex is wild, so the resolve code fetches nonsense for the class (which is not NULL so it's considered resolved).

**Comment by DanHeidinga** [comment]
https://github.com/eclipse/openj9/blob/bdffa781b114e32780a0c73e645b35e87a4e5b41/runtime/vm/resolvesupport.cpp#L256

Try running with `-Xtrace:print=j9vm.142` or formatting the tracefile from the crash to see the cpIndex

**Comment by fjeremic** [comment]
> @NigelYiboYu @fjeremic Does this fail on other platforms as well?

This is a Z specific issue, because the DAA optimization in question only runs on Z. Yes, @NigelYiboYu please see suggestions from Dan and GAC and let's get some more tracing posted here. Appreciate the help guys!

**Comment by NigelYiboYu** [comment]
Sure. I can re-run with the suggested trace option.

I remember that the CPIndex being 70 (0x46) at AOT-Compile time and the relocation didn't touch that. It remained 70 in the vm helper:
`#2  resolveStaticMethodRef (vmStruct=0xbeb2200, ramCP=0xbff3a90, cpIndex=70, resolveFlags=32) at resolvesupport.cpp:588`

I saw the same in GDB live debug.

**Comment by gacholio** [comment]
Perhaps index 70 doesn't represent a class.

If you have a core, use DDR to examine the ROM class.

!j9constantpool 0xbff3a90
find the ramClass field
!j9clas that
find the romClass field
!dumpromclass that

**Comment by gacholio** [comment]
You can also look at the refs directly. If this is 31 bit, RAM CP entries are 8 bytes each, as are ROM CP entries.

!j9constantpool 0xbff3a90
find the romConstantPool field
add 70*8 to the value
!j9romclassref that

For RAM, add 70*8 to 0xbff3a90 and

!j9ramclassref that

**Comment by NigelYiboYu** [comment]
@gacholio Thanks for the debugging suggestions. I used DDR and found something fishy about the AOT patched CP address. Now I think I've found the problem and can fix the JIT to pass my unit test. I'll do more tests to verify my JIT fix later, and will close this issue once things are verified.

The problem is caused by the DAA optimization not setting `inlinedIndex` on BCDCHK nodes. Without correct inline indices, AOT is not able to compute correct relocation CP addresses; and the `# Address Of Constant Pool` fields are patached with (supposedly) garbage values. Bad `ramCP` seems to be the cause of VM side crashes.

The solution is to make DAA set correct inline index on all `BCDCHK` nodes.
